### PR TITLE
Feature: get apiaddress for kubernetes cluster.

### PR DESCRIPTION
**Pull request by yaoyaoio**
add: get apiaddress for kubernetes cluster.

I tested it on aliyun ack cluster.

CleanShot 2023-02-21 at 18 59 46@2x

### CODE REVIEW

File: src/cli/pkg/kubernetes/common.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+func GetStarshipService(client *k8s.Clientset) (*v1.Service, error) {
+	return GetService(client, DefaultNamespace, DefaultServiceName)
+}
+
+func getStarshipServiceURL(client *k8s.Clientset) (string, error) {
+	service, err := GetStarshipService(client)
+	if err != nil {
+		return "", err
+	}
+	return service.Spec.ClusterIP, nil
```
Comment (nascentcore-eng) [inline src/cli/pkg/kubernetes/common.go]: Need comments for all these functions

**Review by nascentcore-eng — COMMENTED** [review]
You might need to run `tools/cleanup.sh` on your local repo to fix the build issue

https://github.com/tricorder-observability/starship/blob/main/README.md#contributing
(I just added after seeing the CI test failures, sorry for the negligence)

**Comment by nascentcore-eng** [comment]
Take a look at the CI test failures, and see if you could fix them.

**Comment by nascentcore-eng** [comment]
Run tools/cleanup.sh to fix some of the style issues and bazel build files issue.

The below are the failures in the CI pipeline.
image

**Comment by yaoyaoio** [comment]
ok. I fix these issues.

**Comment by yaoyaoio** [comment]
I have made changes according suggestions Please review and CI test.

File: src/cli/cmd/module/module.go [hunk 2]
```diff
@@ -16,6 +18,17 @@ var ModuleCmd = &cobra.Command{
 	Run: func(cmd *cobra.Command, args []string) {
 		fmt.Println("module called")
 	},
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		// if starship apiAddress is not set, try to get it from kubernetes
+		if apiAddress == "" {
+			newApiAddress, err := kubernetes.GetStarshipAPIAddress()
+			if err != nil {
+				fmt.Println("connect to kubernetes failed, please use --api-address to set api address manually.")
```
Comment (nascentcore-eng) [inline src/cli/cmd/module/module.go]: If this happens, what's the exit code of the command?
I am not familiar with cobra @JaredTan95

File: src/cli/pkg/kubernetes/client.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	configLocal := clientcmd.NewDefaultClientConfigLoadingRules()
+	startConfig, err := configLocal.GetStartingConfig()
+	if err != nil {
+		return nil, err
+	}
+	return clientcmd.NewDefaultClientConfig(*startConfig, nil).ClientConfig()
+}
+
+// GetClientForConfig returns the kubernetes client for the given config.
+func GetClientForConfig(config *rest.Config) (*k8s.Clientset, error) {
```
Comment (nascentcore-eng) [inline src/cli/pkg/kubernetes/client.go]: This does not seem useful to be a independent helper function.
You can just directly call k8s.NewForConfig(config) in the caller of GetClientForConfig inside NewClient().

File: src/cli/pkg/kubernetes/client.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import (
+	"fmt"
+
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// GetConfig returns the kubernetes config.
+func GetConfig() (*rest.Config, error) {
```
Comment (nascentcore-eng) [inline src/cli/pkg/kubernetes/client.go]: This should be private function

File: src/cli/pkg/kubernetes/common.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	matav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+)
+
+var (
```
Comment (nascentcore-eng) [inline src/cli/pkg/kubernetes/common.go]: nit: const

File: src/cli/pkg/kubernetes/common.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+		}
+	}
+	return service.Spec.Ports[0].Port, nil
+}
+
+// GetStarshipAPIAddress returns the starship Service ClusterIP and serverhttp port
+func GetStarshipAPIAddress() (string, error) {
+	client, err := NewClient()
+	if err != nil {
+		return "", err
```
Comment (nascentcore-eng) [inline src/cli/pkg/kubernetes/common.go]: Use src/utils/errors.Wrap() to return contextual error message, see
https://github.com/tricorder-observability/starship/blob/05ed0bace56f58a49c7897e5ccf31acc76793d45/src/utils/errors/wrap_test.go#L30
for example,

File: src/cli/pkg/kubernetes/common.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	k8s "k8s.io/client-go/kubernetes"
+)
+
+var (
+	DefaultNamespace   = "tricorder"
+	DefaultServiceName = "api-server"
+)
+
+// GetService returns a Service by name and namespace
+func GetService(client *k8s.Clientset, namespace, serviceName string) (*v1.Service, error) {
```
Comment (nascentcore-eng) [inline src/cli/pkg/kubernetes/common.go]: most of these function should be private, adjust them to private if not used outside of this package.

**Review by nascentcore-eng — CHANGES_REQUESTED** [review]

File: src/cli/cmd/module/module.go [hunk 8]
```diff
@@ -27,8 +40,7 @@ var (
 func init() {
 	// Here you will define your flags and configuration settings.
 	ModuleCmd.PersistentFlags().StringVar(&apiAddress, "api-address", "", "address of starship api server.")
-	_ = ModuleCmd.MarkPersistentFlagRequired("api-address")
-
+	//_ = ModuleCmd.MarkPersistentFlagRequired("api-address")
```
Comment (oojimmy) [inline src/cli/cmd/module/module.go]: need to delete this,  not just annotation it.

File: src/cli/pkg/kubernetes/common.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// getStarshipServicePort returns the starship Service serverhttp port
+func getStarshipServicePort(client *k8s.Clientset) (int32, error) {
+	service, err := GetStarshipService(client)
+	if err != nil {
+		return 0, err
+	}
+	for _, port := range service.Spec.Ports {
+		if port.Name == "serverhttp" {
```
Comment (oojimmy) [inline src/cli/pkg/kubernetes/common.go]: "serverhttp" can be set by `const var` in the module, like `	DefaultNamespace   = "tricorder"`

**Comment by nascentcore-eng** [comment]
The go file lacks license header.
You'll need to run tools/cleanup.sh to add the license header

**Comment by nascentcore-eng** [comment]
Starship's CI/CD setup on GitHub might prevent you from releasing a dev image, we'll fix them separately.

**Comment by nascentcore-eng** [comment]
I think my last request is to think about adding an automated test.

I could not come up a plan to implement a simple automated test, because there is no readily available testing fixture (like a fake kubernetes and kubectl local environment).

But I think you can write a brief doc on how to verify the correctness of this feature.

**Comment by JaredTan95** [comment]
> Starship's CI/CD setup on GitHub might prevent you from releasing a dev image, we'll fix them separately.

https://github.com/tricorder-observability/starship/pull/99 fixed docker build and push issue, you can rebase `main` branch to fix CI. @yaoyaoio

**Comment by yaoyaoio** [comment]
> implement

Do you want me to implement a simple automated test to verify getting starship apiserver from kubernetes?

**Comment by nascentcore-eng** [comment]
> 

Yes, if it's possible to do that.

If it's very complicated to do, then don't bother, manual testing is fine.

**Comment by yaoyaoio** [comment]
Sorry I should’t have used merge.

**Comment by nascentcore-eng** [comment]
Need to rebase on the main branch.

File: src/cli/cmd/module/module.go [hunk 10]
```diff
@@ -17,8 +17,11 @@ package module
 
 import (
 	"fmt"
+	"log"
```
Comment (nascentcore-eng) [inline src/cli/cmd/module/module.go]: Use `"github.com/tricorder/src/utils/log"`
It's a drop-in replacement of "log"

File: src/cli/cmd/module/module.go [hunk 11]
```diff
@@ -31,6 +34,16 @@ var ModuleCmd = &cobra.Command{
 	Run: func(cmd *cobra.Command, args []string) {
 		fmt.Println("module called")
 	},
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		// if Starship apiAddress is not set, try to get it from kubernetes
+		if apiAddress == "" {
+			newApiAddress, err := kubernetes.GetStarshipAPIAddress()
+			if err != nil {
+				log.Fatal("connect to kubernetes failed, please use --api-address to set api address manually.")
```
Comment (nascentcore-eng) [inline src/cli/cmd/module/module.go]: `connect to kubernetes failed`
change to
`Failed to connect to Kubernetes API Server`

**Review by nascentcore-eng — APPROVED** [review]
Some minor suggestions

**Pull request merged**